### DISCUSSION TITLE
Find a specific pixel, based on the color

### DISCUSSION THREAD

**Discussion by jmspaggi** [ImageMagick/ImageMagick#2662] [Help] [answered]
Hi,

I need to find the first black pixel for a specific column on a picture.

Today I'm doing something like that:
```./adb exec-out screencap -p > screen.png
convert screen.png -crop 1x1000+100+0 +repage cropped.png
total=999999
line=215

# Search for the first black pixel.
while [ $total -gt 10 ]; do
    colors=`convert cropped.png -crop "1x1+0+$line" txt:- 2>&1 | grep -v pixel | cut -d"(" -f3 | cut -d")" -f1`
    r=`echo $colors | cut -d"," -f1`
    g=`echo $colors | cut -d"," -f2`
    b=`echo $colors | cut -d"," -f3`
    #echo $colors
    total=$((a + b + c))
    line=$[$line+1]
done

echo First black pixel at $line

```

But as you can imagine, it's super slow! I'm wondering if there is something like:
```
convert screen.png -geometry 1x1000+100+0 -color #000000 -threshold 10 -findfirst
```

Where I can define in which rectangle of the picture to search, which color to search for and the threshold. Which means that #030303 will still be considered as black, but #0B0B0B will not. And it returns the coordinates of that non black pixel. 

On optimization I can do is to use `convert cropped.png -crop "1x100+0+1" txt:-` instead of iterating, and iterate on the output. Wondering if there is even "nicer" option.

Thanks,

JMS

**Comment by snibgo** [comment]
What version of IM, on what platform? If v7, I suggest using "magick", not "convert".

We can find the lowest (or highest) pixel values within an image (which could be a crop). For example:
```
magick \
  toes.png -colorspace Gray \
  -define identify:locate=minimum -define identify:limit=1 info:

Channel minimum locations:
  Gray: 6050.94 (0.0923314) 198,207

```
See https://imagemagick.org/script/defines.php and search for "Identify Defines".

This is fast, much faster than a shell loop.

If you want a range of values to be considered "black", then insert a "-threshold" operation.

So your command would be something like:
```
magick \
  screen.png \
  -crop 1x1000+100+0 +repage \
  -threshold 20% \
  -define identify:locate=minimum -define identify:limit=1 info:
```

**Reply by jmspaggi** [reply]
I'm using version 6. That's the one packaged with Debian :-/

```
Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
```

Since I don't have the `magick` command in 6.9, by what should I replace it? `convert`?

Running this command `convert screen.png -crop 1x1000+100+0 +repage -threshold 20% -define identify:locate=minimum -define identify:limit=1 info:` I'm getting this output:
```
convert screen.png -crop 1x1000+100+0 +repage -threshold 20% -define identify:locate=minimum -define identify:limit=1 info:
  Channel minimum locations:
  Red: 0 (0) 0,0
  Green: 0 (0) 0,0
  Blue: 0 (0) 0,0
  Alpha: 0 (0) 0,0
```

What I'm looking for is the first non black pixel after the first black one. 

My full ugly bash script is the following:
```
convert screen.png -crop 1x1000+100+0 +repage cropped.png
total=999999
line=215

# Search for the first black pixel.
while [ $total -gt 10 ]; do
    colors=`convert cropped.png -crop "1x1+0+$line" txt:- 2>&1 | grep -v pixel | cut -d"(" -f3 | cut -d")" -f1`
    r=`echo $colors | cut -d"," -f1`
    g=`echo $colors | cut -d"," -f2`
    b=`echo $colors | cut -d"," -f3`
    #echo $colors
    total=$((a + b + c))
    line=$[$line+1]
done

echo First black pixel at $line

# Now search for the first non-black pixel
while [ $total -lt 10 ]; do
    colors=`convert cropped.png -crop "1x1+0+$line" txt:- 2>&1 | grep -v pixel | cut -d"(" -f3 | cut -d")" -f1`
    r=`echo $colors | cut -d"," -f1`
    g=`echo $colors | cut -d"," -f2`
    b=`echo $colors | cut -d"," -f3`
    #echo $colors
    total=$((a + b + c))
    #echo $total
    line=$[$line+1]
done

echo First non-black pixel at $line
```

I looked at the `define` documentation. Seems to do a LOT! I will have to spend more time to understand it.

**Comment by snibgo** [comment]
Please show your input screen.png.

**Reply by jmspaggi** [reply]
Here you go. When I'm trying to (quickly) find is the beginning of the 2nd box.

![screen](https://user-images.githubusercontent.com/3017827/94939751-8eae5400-04a0-11eb-8386-635a68e44754.png)

**Comment by snibgo** [comment]
You crop with offsets +100+0, and start searching at y=215. We get the same effect by cropping with offsets +100+215.

```
magick 94939751-8eae5400-04a0-11eb-8386-635a68e44754.png -crop 1x1000+100+215 +repage -colorspace Gray -define identify:locate=minimum -define identify:limit=1 info:

Channel minimum locations:
  Gray: 0 (0) 0,381
  Alpha: 65535 (1) 0,0
```
You can extract the coordinates with "cut" and so on. So the first black pixel is at (0,381) in the crop, so this is (100+0,215+381) = (100,596) in screen.png. 

You said:
> I need to find the first black pixel for a specific column on a picture.

But now you say:
> What I'm looking for is the first non black pixel after the first black one.

Okay, well, a second command can crop from that found coordinate, turn all non-black pixels white ("-fill white +opaque black"), and search for the lightest pixel ("-define identify:locate=maximum -define identify:limit=1").

**Reply by jmspaggi** [reply]
My first goal was indeed only the black pixel, which gives me a line, from where I crop and then search for a non-black, the same way.

The example you are giving above works perfectly for the black one. The option with maximum gives me the last of the lightest instead of the first. Or maybe I missed something. If the first command cuts at 596, I can search for the 2nd with something like this:
```convert screen.png -crop 1x1000+100+597 +repage -colorspace Gray -define identify:locate=maximum  -define identify:limit=1 info:```

But it returns 601 for the line. Which means 601+597, which is to much. Should be around 684 in the picture, so around 87 on the command output?

```
$ convert screen.png -crop 1x1000+100+597 +repage -colorspace Gray -define identify:locate=maximum  -define identify:limit=1 info:
  Channel maximum locations:
  Gray: 18899 (0.28838) 0,601
  Alpha: 0 (0) 0,0
```
Or I am doing some maths wrong?

**Comment by fmw42** [comment, marked as answer]
`convert screen.png -crop 1x1000+100+0 +repage sparse-color: | tr " " "\n" | grep "srgb(0,0,0)" | head -n 1`

**Reply by jmspaggi** [reply]
Almost work ;)

It returns srgba(0,0,0,1) I think (for me) and they are all on the same line :-/ So I need to cut that first.
`convert screen.png -crop 1x1000+100+0 +repage sparse-color: | tr ' ' '\n' | tail -n +250 | grep "srgba(0,0,0,1)" | head -n 1`
gives me 596, like above.

Then I can re-use the same to get the next non-black:
`convert screen.png -crop 1x1000+100+0 +repage sparse-color: | tr ' ' '\n' | tail -n +597 | grep -v "srgba(0,0,0,1)" | head -n 1`
Where I get 673. Not exactly correct, it'S a 2,2,2 in the middle of all the blacks.

This: `convert screen.png -crop 1x1000+100+0 +repage sparse-color: | tr ' ' '\n' | tail -n +597 | grep -v "srgba(0,0,0,1)" | grep -v "srgba(1,1,1,1)" | grep -v "srgba(2,2,2,1)" | head -n 1` gives me the 684 that I need.

I like this options too.

**Comment by snibgo** [comment]
I said:
> turn all non-black pixels white ("-fill white +opaque black")

... but you didn't do that. Why not?
```
magick 94939751-8eae5400-04a0-11eb-8386-635a68e44754.png -crop 1x1000+100+597 +repage -colorspace Gray -fill white +opaque black -define identify:locate=maximum -define identify:limit=1 info:

Channel maximum locations:
  Gray: 65535 (1) 0,76
  Alpha: 65535 (1) 0,0
```

**Reply by jmspaggi** [reply]
Ha! Sorry! My bad! I went to fast on that part and forgot to do it :-/ 76+597 = 673 which is my 2,2,2 guy close to what I need. I think I can definitely make that work! Thanks a lot all for your help.
